src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingEndpoint.cs

```diff
@@ -147,6 +184,120 @@ public async Task<TextEditContainer> Handle(DocumentRangeFormattingParams reques
+                return null;
+            }
+
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            var hostDocumentIndex = request.Position.GetAbsoluteIndex(sourceText);
+            var triggerCharacterKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            if (triggerCharacterKind is not RazorLanguageKind.CSharp)
+            {
+                _logger.LogInformation($"Unsupported trigger character language {triggerCharacterKind:G}.");
+                return null;
+            }
+
+            if (!IsApplicableTriggerCharacter(request.Character, triggerCharacterKind))
+            {
+                // We were triggered but the trigger character doesn't make sense for the current cursor position. Bail.
+                _logger.LogInformation("Unsupported trigger character location.");
+                return null;
+            }
+
+            if (!_razorDocumentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out _, out var projectedIndex))
+            {
+                _logger.LogWarning($"Failed to map to projected position for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var formattingContext = FormattingContext.Create(
+                request.TextDocument.Uri, documentSnapshot, codeDocument, request.Options, _adhocWorkspaceFactory, isFormatOnType: true);
+            var documentOptions = await GetDocumentOptionsAsync(request, formattingContext.CSharpWorkspaceDocument).ConfigureAwait(false);
+
+            // Ask C# for formatting changes.
+            var formattingChanges = await RazorCSharpFormattingInteractionService.GetFormattingChangesAsync(
+                formattingContext.CSharpWorkspaceDocument, typedChar: request.Character[0], projectedIndex, documentOptions, cancellationToken).ConfigureAwait(false);
+
+            formattingContext.Dispose();
```

Comment:
Would it be possible to do `using var formattingContext ...` above or would disposing it at the end of the method have other unintended consequences?

Reply:
No, I think you're right! I somehow forgot about the `using` convention.